src/arch/riscv/pma_checker.cc

```diff
@@ -94,6 +118,42 @@ PMAChecker::takeOverFrom(BasePMAChecker *old)
     PMAChecker* derived_old = dynamic_cast<PMAChecker*>(old);
     assert(derived_old != nullptr);
     uncacheable = derived_old->uncacheable;
+    misaligned = derived_old->misaligned;
+}
+
+Fault
+PMAChecker::checkPAddrAlignment(
+    const RequestPtr &req, BaseMMU::Mode mode, Addr vaddr)
+{
+    Addr addr = 0;
+    if (mode != BaseMMU::Execute) {
+      assert(req->hasPaddr());
+      addr = req->getPaddr();
+      Addr alignSize = mask(req->getArchFlags() & 0x7) + 1;
+      auto it = misaligned.contains(RangeSize(addr, req->getSize()));
+      if (it == misaligned.end()) {
+          return checkAlignmentInternal(
```

Comment:
I think this is not correct.

From my understanding around the discussion on [the previous change](https://gem5-review.googlesource.com/c/public/gem5/+/46279/comments/084355fe_df1c3950), the `vaddr` is only used as the value that is returned in the fault. It should be be used as the address we're checking.

For instance, on the PTW, we need to check for alignment on the physical address, but raise a fault which has the virtual address.

I *think* this code is actually checking the `vaddr` for alignment, not the physical address that the PTW is using.

Reply:
Thanks @powerjg point out the issue of checking physicall address alignment. I have updated the patch to fix the issue.